tools/ideabuck/src/com/facebook/buck/intellij/ideabuck/run/BuckBuildBeforeRun.java

```diff
@@ -0,0 +1,170 @@
+  public Icon getIcon() {
+    return BuckIcons.BUCK_TOOL_WINDOW_ICON;
+  }
+
+  @Override
+  public boolean canExecuteTask(
+      RunConfiguration configuration, ExternalSystemBeforeRunTask beforeRunTask
+  ) {
+    if (!(configuration instanceof ApplicationConfiguration)) {
+      return false;
+    }
+    VirtualFile containingFile = ((ApplicationConfiguration) configuration).getMainClass().getContainingFile().getVirtualFile();
+    VirtualFile buckFile = BuckFileUtil.getBuckFile(containingFile);
+
+    return buckFile != null;
+  }
+
+  @Override
+  public boolean executeTask(
+      final DataContext context,
+      RunConfiguration configuration,
+      ExecutionEnvironment env,
+      ExternalSystemBeforeRunTask beforeRunTask
+  ) {
+    if (currentProject == null) {
+      currentProject = configuration.getProject();
+    }
+
+    if (configuration instanceof ApplicationConfiguration) {
+      VirtualFile containingFile = ((ApplicationConfiguration) configuration).getMainClass()
+          .getContainingFile().getVirtualFile();
+
+      VirtualFile buckFile = BuckFileUtil.getBuckFile(containingFile);
+
+      if (buckFile == null) {
+        return false;
+      }
+
+      BuckBuildManager buildManager = BuckBuildManager.getInstance(currentProject);
+      BuckQueryCommandHandler buckQueryCommandHandler =
+          new BuckQueryCommandHandler(
+              currentProject,
+              buckFile.getParent(),
+              BuckCommand.QUERY,
+              new Function<List<String>, Void>() {
+                @javax.annotation.Nullable
+                @Override
+                public Void apply(@javax.annotation.Nullable List<String> strings) {
+                  if (strings == null
+                      || strings.isEmpty()
+                      || strings.get(0) == null
+                      || strings.get(0).isEmpty()) {
+                    return null;
+                  }
+                  String targetResultCandidate = strings.get(0);
+                  for (String candidate : strings) {
+                    if (!candidate.contains("lint")) {
+                      targetResultCandidate = candidate;
+                    }
+                  }
+                  targetResult = targetResultCandidate;
+                  return null;
+                }
+              });
+      buckQueryCommandHandler.command().addParameter("owner(" + containingFile.getPath() + ")");
+      buildManager.runInCurrentThreadPostEnd(
+          buckQueryCommandHandler,
+          () -> {
+            BuckBuildCommandHandler buckBuildCommandHandler =
+                new BuckBuildCommandHandler(
+                    currentProject,
+                    buckFile.getParent(),
+                    BuckCommand.BUILD,
+                    true
+                );
+            if (targetResult != null) {
+              buckBuildCommandHandler.command()
+                  .addParameters("--deep", targetResult, "--build-report", "buck-out/out.json");
```

Comment:
why `--deep`?

Also, I'd recommend avoid writing to buck-out, in this case a temporary folder would just fine.

Reply:
All of the resulting jars from all of the transitive dependencies has to be present to be put on the runtime classpath.